### PR TITLE
New pruning mechanism

### PR DESCRIPTION
**Pull request by twaugh**
This prunes builds, including those without associated BuildConfig objects.

It does not prune images or configmaps.

### CODE REVIEW

**Comment by twaugh** [comment]
Ready for review now.

File: README.md [hunk 1]
```diff
@@ -126,11 +126,22 @@ Role Variables
...
+    - adm
+    - prune
+    - builds
+    - --orphans=true
+    - --confirm
+    osbs_pruner_schedule_build: "0 0 * * *"
+    # Automatically prunes pods from CronJobs
+    osbs_pruner_successful_jobs: 5
+    osbs_pruner_failed_jobs: 5
+    # Define this variable to enable pruning: will be created if absent
```
Comment (lcarva) [inline README.md]: What will be created?

File: defaults/main.yml [hunk 2]
```diff
@@ -46,17 +46,24 @@ osbs_buildroot_imagestream: ''
...
-
 osbs_odcs_enabled: false
 osbs_odcs_signing_intents: {}
 osbs_odcs_default_signing_intent: null
 osbs_odcs_api_url: ''
 osbs_odcs_auth_ssl_certs_dir: /usr/share/osbs
 
 osbs_reactor_config_maps: []
+
+osbs_pruner_template_dest: /tmp
```
Comment (lcarva) [inline defaults/main.yml]: Can we use `osbs_openshift_home` instead?

File: tasks/pruner.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
+- name: Set Facts
+  set_fact:
+      osbs_build_pruner_cronjob_yaml: "{{ osbs_pruner_template_dest }}/build-pruner-cronjob.yaml"
+
+- name: Query pruner service account
```
Comment (lcarva) [inline tasks/pruner.yml]: For service account and clusterrolebinding manipulation, can we use the same pattern that is done in `tasks/main.yml`? It's a little more resilient.

File: templates/build-pruner-cronjob.yaml.j2 [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+metadata:
+  name: build-pruner
+spec:
+  schedule: "{{ osbs_pruner_schedule_build }}"
+  concurrencyPolicy: Forbid
+  successfulJobsHistoryLimit: {{ osbs_pruner_successful_jobs }}
+  failedJobsHistoryLimit: {{ osbs_pruner_failed_jobs }}
+  jobTemplate:
+    spec:
+      template:
```
Comment (lcarva) [inline templates/build-pruner-cronjob.yaml.j2]: Adding `metadata.labels.parent` might make it easier to manage these pods.

Comment (lcarva) [reply templates/build-pruner-cronjob.yaml.j2]: Ok, fair enough. I think it's fine as long as we can query for them.

File: README.md (hunk 1 above)
Comment (rcerven) [reply README.md]: pruner service account ?

File: tasks/main.yml [hunk 5]
```diff
@@ -190,25 +162,8 @@
...
-  environment: "{{ osbs_environment }}"
-  when: osbs_prune and yaml_cronjob.changed
-  tags:
-  - oc
-
 - include: orchestrator.yml
   when: osbs_orchestrator
+
+- include: pruner.yml
+  when: "osbs_pruner_serviceaccount is defined"
```
Comment (rcerven) [inline tasks/main.yml]: if you will make it dependent on existence of service account how will it create it when it doesn't exists?

File: README.md (hunk 1 above)
Comment (twaugh) [reply README.md]: I'll clarify as "# Define this variable to enable pruning: account will be created if absent"

File: defaults/main.yml (hunk 2 above)
Comment (twaugh) [reply defaults/main.yml]: Yes, updated.

File: tasks/pruner.yml (hunk 3 above)
Comment (twaugh) [reply tasks/pruner.yml]: Yes, and for the CronJob. Updated.

File: README.md [hunk 6]
```diff
@@ -126,11 +126,22 @@ Role Variables
...
+    - --orphans=true
+    - --confirm
+    osbs_pruner_schedule_build: "0 0 * * *"
+    # Automatically prunes pods from CronJobs
+    osbs_pruner_successful_jobs: 5
+    osbs_pruner_failed_jobs: 5
+    # Define this variable to enable pruning: account will be created if absent
+    osbs_pruner_serviceaccount: pruner
+    # Which cluster-role to grant to the service account
+    osbs_pruner_clusterrole_build: system:openshift:controller:build-controller
```
Comment (lcarva) [inline README.md]: nice!

File: tasks/pruner.yml [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  - oc
+
+- name: Copy cluster rolebinding for pruner service account
+  template:
+    src: "openshift-rolebinding.{{ item.yaml_version | default('v2') }}.yml.j2"
+    dest: "{{ osbs_openshift_home }}/{{ inventory_hostname }}-{{ osbs_namespace }}-rolebinding-{{ item.name }}.yml"
+  with_items:
+  - name: build-pruner
+    role: "{{ osbs_pruner_clusterrole_build }}"
+    yaml_version: v1
```
Comment (lcarva) [inline tasks/pruner.yml]: Should this be `v2`? the `v1` template doesn't read from `serviceaccounts`.

Maybe remove this line, and hard code `v2` in the `template.src` above?

Comment (twaugh) [reply tasks/pruner.yml]: Ah, will change to `v2`.

The template is shared with tasks in main.yml, so I don't think I can change it.

Comment (lcarva) [reply tasks/pruner.yml]: I don't mean change the actual template, but change the template key in this task to:
```
src: "openshift-rolebinding.v2.yml.j2"
```

Comment (twaugh) [reply tasks/pruner.yml]: Gotcha. Fixed.

**Pull request merged**